compute_tools/src/installed_extensions.rs

```diff
@@ -0,0 +1,227 @@
+                // update the highest_version if new version is higher
+                // update the lowest_version if new version is lower
+                let version = v.to_string();
+                extensions_map
+                    .entry(extname.to_string())
+                    .and_modify(|e| {
+                        // convert version to SemanticVersion
+                        let version_sem =
+                            SemanticVersion::from_str(&version).expect("failed to parse version");
+                        // use SemanticVersion to compare versions
+                        let lowest_version_sem = SemanticVersion::from_str(&e.lowest_version)
+                            .expect("failed to parse lowest version");
+                        let highest_version_sem = SemanticVersion::from_str(&e.highest_version)
+                            .expect("failed to parse highest version");
+
+                        debug!(
+                            "extname: {}, version: {}, lowest: {}, highest: {}",
+                            extname, version, e.lowest_version, e.highest_version
+                        );
+
+                        if lowest_version_sem > version_sem {
+                            e.lowest_version = version.clone();
+                        }
+                        if highest_version_sem < version_sem {
+                            e.highest_version = version.clone();
+                        }
+
+                        // count the number of databases where the extension is installed
+                        e.n_databases += 1;
+                    })
+                    .or_insert(InstalledExtension {
+                        extname: extname.to_string(),
+                        lowest_version: version.clone(),
+                        highest_version: version.clone(),
+                        n_databases: 1,
+                    });
+            }
+        }
+
+        Ok(InstalledExtenstions {
+            extensions: extensions_map.values().cloned().collect(),
+        })
+    })
+    .await?
+}
+
+pub fn log_installed_extensions(connstr: Url) -> Result<()> {
+    let connstr = connstr.clone();
+
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .expect("failed to create runtime");
+    let result = rt
+        .block_on(get_installed_extensions(connstr))
+        .expect("failed to get installed extensions");
+
+    info!(
+        "[INSTALLED_EXTENSIONS]: {}",
+        serde_json::to_string(&result).expect("failed to serialize extensions list")
+    );
+    Ok(())
+}
+
+//-------------------------------------------------------------------------
+//
+// Most of the postgres extensions use 2 part versioning (major.minor)
+// Some extensions use 3 part versioning (major.minor.patch)
+//
+
+#[derive(Debug, Clone, Copy)]
+struct SemanticVersion {
+    major: u32,
+    minor: u32,
+    patch: Option<u32>, // Make patch optional
```

Comment:
Postgres is totally OK with versions like `1.4a`
```
postgres=# select oid, extname, extversion from pg_extension;
  oid  | extname | extversion 
-------+---------+------------
 13872 | plpgsql | 1.0
 16384 | neon    | 1.4a
(2 rows)
```

Even in the semver standard at least patch shouldn't be necessary an int https://semver.org/

In this PR we deal with third-party extensions, so we don't really want to pay attention to how they are versioned. Yes, most of them will be just, maybe all, but we don't know and we don't know to check it (there could be some non-int version in-between min/latest some use could install)

I think there are three ways to deal with it
1. Store raw string inside SemanticVersion (+change ints to be optional) and fallback to lexicographic sorting if parsing failed
2. Just print a warning, ignore such extensions, and proceed further -> I assume that it's better to have some information than nothing at all
3. Error out gracefully and stop -> minimal effort

I'd probably prefer 1., shouldn't be hard, easy to write more test cases and should be robust

Reply:
After writing this, I realized that we already have https://docs.rs/semver/latest/semver/ in the `Cargo.lock`. It's some indirect dependency of `rustc_version` and up the stack, so it's probably fine to add it explicitly or maybe we can just use it (I'd double check with some Rust folks)

It's probably better than writing some custom code

Postgres is anyway not fully semver-compliant, so it won't protect us from versions like `1.4a` in the example above. And just for safety I'd wrap it into our own wrapper struct that falls back to lexicographic sorting

Reply:
fixed

Reply:
I'd suggest returning a full list of all versions that are present in any databases. If a cluster has a mix of different versions in different databases, it'd be nice to know which versions they are. And then you don't need the comparison functions for version numbers at all.

Reply:
do you think we need to map them to databases, where the extensions are installed, or just return array of versions that are present?

Reply:
Just returning an array should be enough